pkg/controller/networkpolicy/controller_query_perf_test.go

```diff
@@ -0,0 +1,145 @@
+	v1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"sync"
+	"testing"
+	"time"
+)
+
+/*
+TestLargeScaleEndpointQueryManyPolicies tests the execution time and the memory usage of computing a scale
+of 10k Namespaces, 10k NetworkPolicies, 10k Pods, where query returns every policy (applied + ingress).
+
+NAMESPACES   PODS    NETWORK-POLICIES    TIME(s)    MEMORY(M)
+1            10000   10000               13.88       1092
+
+The metrics are not accurate under the race detector, and will be skipped when testing with "-race".
+*/
+func TestLargeScaleEndpointQueryManyPolicies(t *testing.T) {
+	namespace := rand.String(8)
+	getObjects := func() ([]*v1.Namespace, []*networkingv1.NetworkPolicy, []*v1.Pod) {
+		namespaces := []*v1.Namespace{
+			{
+				ObjectMeta: metav1.ObjectMeta{Name: namespace, Labels: map[string]string{"app": namespace}},
+			},
+		}
+		uid := rand.String(8)
+		networkPolicies := []*networkingv1.NetworkPolicy{
+			{
+				ObjectMeta: metav1.ObjectMeta{Namespace: namespace, Name: "np-1" + uid, UID: types.UID(uuid.New().String())},
+				Spec: networkingv1.NetworkPolicySpec{
+					PodSelector: metav1.LabelSelector{MatchLabels: map[string]string{"app-1": "scale-1"}},
+					PolicyTypes: []networkingv1.PolicyType{networkingv1.PolicyTypeIngress, networkingv1.PolicyTypeEgress},
+					Ingress: []networkingv1.NetworkPolicyIngressRule{
+						{
+							From: []networkingv1.NetworkPolicyPeer{
+								{
+									PodSelector: &metav1.LabelSelector{
+										MatchLabels: map[string]string{"app-1": "scale-1"},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+		pods := []*v1.Pod{
+			{
+				ObjectMeta: metav1.ObjectMeta{Namespace: namespace, Name: "pod1" + uid, UID: types.UID(uuid.New().String()), Labels: map[string]string{"app-1": "scale-1"}},
+				Spec:       v1.PodSpec{NodeName: getRandomNodeName()},
+				Status:     v1.PodStatus{PodIP: getRandomIP()},
+			},
+		}
+		return namespaces, networkPolicies, pods
+	}
+	namespaces, networkPolicies, pods := getXObjects(10000, getObjects)
+	testQueryEndpoint(t, 25*time.Second, namespaces[0:1], networkPolicies, pods, 10000)
+}
+
+func testQueryEndpoint(t *testing.T, maxExecutionTime time.Duration, namespaces []*v1.Namespace, networkPolicies []*networkingv1.NetworkPolicy, pods []*v1.Pod, responseLength int) {
+	// Stat the maximum heap allocation.
+	var wg sync.WaitGroup
+	stopCh := make(chan struct{})
+	var maxAlloc uint64
+	wg.Add(1)
+	go func() {
+		statMaxMemAlloc(&maxAlloc, 500*time.Millisecond, stopCh)
+		wg.Done()
+	}()
+	// create controller
+	objs := toRunTimeObjects(namespaces, networkPolicies, pods)
+	_, querier := makeControllerAndEndpointQueryReplier(objs...)
+	// Everything is ready, now start timing.
+	start := time.Now()
+	// track execution time by calling query endpoint 1000 times on random pods
```

Comment:
s/pods/Pods

Reply:
what do you mean?

Reply:
that's a shortcut (sed syntax) for: replace the "pods" spelling with "Pods"